### PR TITLE
feat(dropdown): added basic dropdown

### PR DESCRIPTION
**Pull request by endritbajo**
Reviewers: @endritbajo @spirosikmd @katia @kdabo @NurahSadek @beerecca @keepitterron @dine 

Added basic dropdown and upgraded stylabilla to the latest version.

#### Accessibility
Can open the menu by pressing `enter` on the dropdown button.
Can navigate the menu with `tab` or arrow `up`/`down`.
Can close the menu with `esc`.
Can select the item by pressing `enter` on the menu item.

dropdown

### CODE REVIEW

File: src/Dropdown/Readme.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      {text: 'Choice 1', value: '1'},
+      {text: 'Choice 2', value: '2'},
+      {text: 'Choice 3', value: '3'}
+    ]}/>
+  <br/>
+  <br/>
+  <br/>
+  <br/>
+  <br/>
+  <br/>
```
Comment (endritbajo) [inline src/Dropdown/Readme.md]: Hack to adjust the height of the example container. We currently don't use a custom template for the styleguide.  I created a ticket for that (STYLE-58).

File: src/Dropdown/index.js [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+    this.state = {
+      menuButtonText: placeholder || textOfFirstMenuItem,
+      isOpen: false,
+    };
+  }
+
+  selectItem(item) {
+    if (this.props.onItemSelect) {
+      this.props.onItemSelect([item]);
```
Comment (endritbajo) [inline src/Dropdown/index.js]: We're returning the selected item as an array because the next dropdowns we will have the option to select multiple items.

File: src/Dropdown/index.js [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    e.nativeEvent.stopImmediatePropagation();
+    this.selectItem(item);
+    this.closeMenuAndFocusMenuButton();
+  }
+
+  handleMenuItemKeyDown(item, itemIndex, e) {
+    e.preventDefault();
+
+    switch (e.keyCode) {
+      case 32: // space
```
Comment (kdabo) [inline src/Dropdown/index.js]: Which keyboard shortcuts should we use to navigate? Is this something to check with Product and Design?

File: src/Dropdown/index.js [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        this.closeMenuAndFocusMenuButton();
+        break;
+      case 38: // arrow up
+        this.focusMenuItemAt(itemIndex - 1);
+        break;
+      case 9: // tab
+      case 40: // arrow down
+        this.focusMenuItemAt(itemIndex + 1);
+        break;
+    }
```
Comment (keepitterron) [inline src/Dropdown/index.js]: shouldn't we mimic a html select's behaviour?
difference being: enter\tab do nothing and up/down when focused open the menu.

Comment (keepitterron) [inline src/Dropdown/index.js]: also, button down scrolls the page together with the elements

File: src/Dropdown/index.js [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+Dropdown.propTypes = {
+  /** Any extra modifier classes you want on the outer form control. */
+  className: PropTypes.string,
+  /** Displays only if validation is set to `error`. */
+  errorMessage: PropTypes.string,
+  /** Used for `id` on the button and `for` on the label. */
+  id: PropTypes.string.isRequired,
+  /** Items to display in the in the dropdown menu. */
```
Comment (keepitterron) [inline src/Dropdown/index.js]: `in the in the`

File: src/Dropdown/index.js [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    this.handleMenuButtonClick = this.handleMenuButtonClick.bind(this);
+    this.handleMenuItemClick = this.handleMenuItemClick.bind(this);
+    this.handleMenuItemKeyDown = this.handleMenuItemKeyDown.bind(this);
+    this.closeMenu = this.closeMenu.bind(this);
+
+    const { items, placeholder } = this.props;
+    const textOfFirstMenuItem = items.length > 0 ? items[0].text : '';
+
+    this.state = {
+      menuButtonText: placeholder || textOfFirstMenuItem,
```
Comment (keepitterron) [inline src/Dropdown/index.js]: isn't it weird that the first element is always selected?

File: src/Dropdown/index.js [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+  focusMenuItemAt(index) {
+    if (index >= 0 && index < this.menuItemElements.length) {
+      this.menuItemElements[index].focus();
+    }
+  }
+
+  openMenu() {
+    this.setState({ isOpen: true });
+    this.addClickOutListener();
```
Comment (keepitterron) [inline src/Dropdown/index.js]: 🎉

File: styleguide.config.js [hunk 8]
```diff
@@ -10,6 +10,10 @@ const webpackConfig = {
         test: /\.css$/,
         loader: 'style-loader!css-loader',
       },
+      {
+        test: /\.(woff|woff2)$/,
+        loader: 'file-loader',
+      },
```
Comment (keepitterron) [inline styleguide.config.js]: mmm
why is this here?

Comment (endritbajo) [inline styleguide.config.js]: Because stylabilla is importing some fonts and the react-styleguidedist cannot load them :/

File: src/Dropdown/index.js [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+  /** Items to display in the dropdown menu. */
+  items: PropTypes.arrayOf(
+    PropTypes.shape({
+      text: PropTypes.string.isRequired,
+      value: PropTypes.string.isRequired,
+    })
+  ).isRequired,
+  /** Label for the dropdown component. */
+  labelMessage: PropTypes.string.isRequired,
+  /** Menu button text placeholder */
```
Comment (beerecca) [inline src/Dropdown/index.js]: Dropdown rather than menu?

Comment (beerecca) [inline src/Dropdown/index.js]: Placeholder shouldn't be required if you have set some selected items right? We're using react-required-if in the upload component which might be useful?

File: src/Dropdown/index.js [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+  placeholder: PropTypes.string.isRequired,
+  /** Called when a dropdown menu item is selected */
+  onItemSelect: PropTypes.func,
+  /** Indexes of the default selected items */
+  selectedItems: PropTypes.arrayOf(PropTypes.number),
```
Comment (beerecca) [inline src/Dropdown/index.js]: I think this should be the values of the selected items, not the indices. Safer with react, also easier if you are trying to set some values.

File: src/Dropdown/index.js [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+  /** If true, show the label. */
+  showLabel: PropTypes.bool,
+  /** If error, displays error message. */
+  validation: PropTypes.oneOf(['none', 'error']),
+  autoFocus: PropTypes.bool,
```
Comment (beerecca) [inline src/Dropdown/index.js]: Not used anywhere

File: src/Dropdown/index.js [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+ * All props without a description are <a target="_blank" href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Attributes">default html button element attributes</a>.
+ */
+class Dropdown extends React.Component {
+  constructor(props) {
+    super(props);
+    this.handleMenuButtonClick = this.handleMenuButtonClick.bind(this);
+    this.handleMenuButtonKeyDown = this.handleMenuButtonKeyDown.bind(this);
+    this.handleMenuItemClick = this.handleMenuItemClick.bind(this);
+    this.handleMenuItemKeyDown = this.handleMenuItemKeyDown.bind(this);
+
```
Comment (beerecca) [inline src/Dropdown/index.js]: Unnecessary whitespazz

File: src/Dropdown/index.js [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+    this.closeMenu = this.closeMenu.bind(this);
+
+    const selectedItems = this.props.selectedItems ? [this.props.selectedItems[0]] : [];
```
Comment (beerecca) [inline src/Dropdown/index.js]: Why does only the first selected item get set?

File: src/Dropdown/index.js [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+  }
+
+  closeMenu() {
+    this.setState({ isOpen: false });
+    this.resetFocusedItems();
+    this.removeClickOutListener();
+  }
+
+  focusItemAt(index) {
+    if (index !== undefined && index >= 0 && index < this.menuItemElements.length) {
```
Comment (beerecca) [inline src/Dropdown/index.js]: Won't ever be undefined

File: src/Dropdown/index.js [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+        break;
+    }
+  }
+
+  componentWillUnmount() {
+    this.removeClickOutListener();
+  }
+
+  getMenuButtonText() {
+    return get(this.props.items, `${this.state.selectedItems[0]}.text`, this.props.placeholder);
```
Comment (beerecca) [inline src/Dropdown/index.js]: Would prefer to just do this old school rather than import get just for this one case...

File: src/Dropdown/index.js (hunk 13 above)
Comment (endritbajo) [inline src/Dropdown/index.js]: It was because it's a single choice dropdown and there cannot be multiple items selected. I put it as array because when we will have multiple choices (checkbox dropdown) we need to keep track of all the check items.
